File: src/neighborhood.c
/*

  define symmetric neighborhoods for unique values u in x

  input
  =====
  n: the sample size
  nu: number of unique x values 
  cumtabu: n times the cumulative empirical df at u
  cumtabx: n times the cumulative empirical df at x
  tabx: frequency of x
  radius: n times the bandwidth

  output specific to neighborhood's 
  =================================
  first: the first neighbor
  size:  the size the neighborhood
  neighbors sorted from the first to last neighborhood
  
*/

#include <stdlib.h> 
void neighborhood(int *first,
		  int *size,
		  int *cumtabu,
		  int *cumtabx,
		  int *tabx,
		  int *radius,
		  int *nu,
		  int *n){
  int u,last;
  
  for (u=0;u<*nu;u++){
    
    /* make a first guess */
    
    first[u]=cumtabu[u]-*radius;
    last=cumtabu[u]+*radius;
    
    /* if x[first[u]] is tied, move
       to the first[u] member in the bin */
    
    if (first[u]<=0) first[u]=1;
    else first[u] = cumtabx[first[u]-1]-tabx[first[u]-1]+1;
    
    /* if x[last] is tied and not the last
       in its bin, move to the previous bin */
    
    if (last>*n) last=*n;
    else if (cumtabx[last-1] > last) last=cumtabx[last-1]-tabx[last-1];
    
    size[u]=last-first[u]+1;
  }
}

int neworder (int *a, int *b){
  if (*a < *b) return -1; else return 1;}

void neighbors(int *first,
	       int *size,
	       int *orderx,
	       int *neighbors,
	       int *nu){
  int u,i,new,start=0;

  /* fill the neighborhoods */
  new=0;
  for (u=0;u<*nu;u++){
    for (i=0;i<size[u];i++){
      neighbors[new] = orderx[first[u]-1+i];
      new++;
    }

    /*     sort the neighborhood */
    qsort(neighbors+start,size[u],(size_t) sizeof(int),(int (*)(const void *, const void *))(neworder));  

    start+=size[u];
  }
}
  

File: R/neighborhood.R
"neighborhood" <- function(x,bandwidth=NULL,kernel="box"){
  N <- length(x)
  if (N<2) stop("Not enough observations for kernel smoothing.")
  
  orderx <- order(x)
  values <- sort(unique(x))
  NU <- length(values)
  workx <- factor(x,labels=1:NU)
  tabu <- tabulate(workx)
  cumtabu <- cumsum(tabu)
  cumtabx <- rep(cumtabu,tabu)
  tabx <- rep(tabu,tabu)
  if (!length(bandwidth)){ ## need a bandwidth (dpik is from KernSmooth)
    ## require(KernSmooth)
    bandwidth <- dpik(cumtabx/N,kernel="box")
  }
  else
    if (bandwidth=="smooth") bandwidth <- N^{-1/4}
  radius <- floor(bandwidth*N)
  
  nbh <- .C("neighborhood",first=integer(NU),size=integer(NU),as.integer(cumtabu),as.integer(cumtabx),as.integer(tabx),as.integer(radius),as.integer(NU),as.integer(N),PACKAGE="prodlim")
  nall <- sum(nbh$size)
  neighbors <- .C("neighbors",first=nbh$first,size=nbh$size,as.integer(orderx),neighbors=integer(nall),as.integer(NU),PACKAGE="prodlim")$neighbors
  
  out <- list(values=values,
              first.nbh=nbh$first,
              size.nbh=nbh$size,
              neighbors=neighbors,
              bandwidth=bandwidth,
              kernel=kernel,
              nu=NU,
              n=N)
    
  class(out) <- "neighborhood"
  out
}
